proposals/0010-vk-buffer-ref.md

```diff
@@ -0,0 +1,270 @@
+* There is no conversion from uint64_t to vk:buffer_ptr.
+* There is no explicit pointer arithmetic. All addressing is implicitly done using the `.` pointer, or indexing an array in the struct T.
+* The comparison operators == and != are not supported for buffer pointers. Comparison to null pointer should be done by first converting to uint64_t.
+
+Most of these restrictions are there for safety. They minimize the possibility of getting an invalid pointer. If the `.` operator is used on a null pointer, the behaviour is undefined.
+
+When used as a member in a buffer, vk::buffer_ptr can be used to pass physical buffer addresses into a shader, and address and access buffer space with logical addressing, which allows tools such as spirv-opt, spirv-reflect and renderdoc to be able to better work with these shaders.
+
+For example, here is a shader using vk::buffer_ptr to do the same thing as the shader above using vk::RawBufferLoad. Note the natural, logical syntax of the reference:
+
+```c++
+
+struct Globals_s
+{
+      float4 g_vSomeConstantA;
+      float4 g_vTestFloat4;
+      float4 g_vSomeConstantB;
+};
+
+typedef vk::buffer_ptr<Globals_s> Globals_p;
+
+struct TestPushConstant_t
+{
+      Globals_p m_nBufferDeviceAddress;
+};
+
+[[vk::push_constant]] TestPushConstant_t g_PushConstants;
+
+float4 MainPs(void) : SV_Target0
+{
+      float4 vTest = g_PushConstants.m_nBufferDeviceAddress.g_vTestFloat4;
+      return vTest;
+}
+
+```
+
+In SPIR-V, Globals_t would be a pointer to the physical buffer storage class. The type of the push constant would contain one of those pointers. The SPIR-V for this shader can be seen in Appendix B. Note the logical context of the declaration and addressing of underlying struct Globals_t including Offset decorations all Globals_t members.
+
+## Linked Lists and Local Variables
+
+vk::buffer_ptr can be used to program a linked list of identical buffers:
+
+```c++
+
+// Forward declaration
+typedef struct block_s block_t;
+typedef vk::buffer_ptr<block_t> block_p;
+
+struct block_s
+{
+      float4 x;
+      block_p next;
+};
+
+struct TestPushConstant_t
+{
+      block_p root;
+};
+
+[[vk::push_constant]] TestPushConstant_t g_PushConstants;
+
+float4 MainPs(void) : SV_Target0
+{
+      block_p g_p = g_PushConstants.root;
+      g_p = g_p.next;
+      if (uint64_t(g_p) == 0) return float4(0.0,0.0,0.0,0.0);
+      return g_p.x
+}
+
+```
+
+Note also the ability to create local variables of type vk::buffer_ptr such as g_p which can be read, written and dereferenced.
+
+## Design Details
+
+### Differences from C++ Pointers
+
+vk::buffer_ptr is different from a C++ pointer in that a selection “.” operation can and must be applied to a buffer pointer to de-reference it. 
+
+### Buffer Pointer Target Alignment
+
+The target alignment `A` of `vk::buffer_ptr(T,A)` must be at least as large as the largest basic type in the buffer reference struct `T`.
```

Comment:
I’m not sure this is correct. Is the intent here to match C alignment rules?

Reply:
This was just something I deduced based on my understanding of the overall feature.

When you say you think this is not correct, are you thinking:

1) `A` cannot be as small as the largest basic type alignment in the buffer, or

2) `A` can be smaller than the largest basic type alignment in the buffer?

Reply:
On many architectures alignments aren’t strictly tied to the sizes of basic types. For example, what is the required alignment for a float3?

If I have a float3 inside a struct, what is the alignment requirement of the struct?

In C a struct’s alignment is defined as the alignment of the most restrictive member (which is not the size of the member).

Reply:
I think what Greg intended was correct, but is not worded clearly.

Correct me if I am wrong, but I think Greg meant to say:

> The target alignment `A` of `vk::buffer_ptr(T,A)` must be at least as large as the largest **alignment** of a basic type in the buffer reference struct `T`.

I believe this is the same as what Chris mentioned: 

> In C a struct’s alignment is defined as the alignment of the most restrictive member

Reply:
That’s definitely part of it.

Is the expected alignment based on the CPU or GPU data layout rules? The HLSL buffer packing rules don’t match C struct packing rules, so we probably need to be a bit explicit here about what is expected (see: https://github.com/microsoft/DirectXShaderCompiler/wiki/Buffer-Packing)

Reply:
Yes, Steven is right about my intent to refer to alignment.

I think part of the problem is my use of *basic* type. For example, for a float3, I consider the *basic* type is float. Is there a better term here? Elemental?

So, one implication of my statement is: if the buffer_ptr pointed to a struct with float3, the alignment A must be at least 4 bytes. Otherwise the compiler would give an error.

Practically, the way I use A is that I take alignment that the compiler gives me for the buffer member given the packing that it has applied to the buffer, then when I generate the SPIR-V OpLoad I use the minimum of that alignment and A.

Does this clear up all concerns?

Reply:
I think what you're saying makes sense, but we need to get language into the spec that reflects the required alignment constraints.